src/python/pysss_nss_idmap.c

```diff
@@ -582,16 +583,24 @@ initpysss_nss_idmap(void)
     if (module == NULL)
         MODINITERROR;
 
-    PyModule_AddIntConstant(module, "ID_NOT_SPECIFIED",
+    ret = PyModule_AddIntConstant(module, "ID_NOT_SPECIFIED",
                             SSS_ID_TYPE_NOT_SPECIFIED);
-    PyModule_AddIntConstant(module, "ID_USER", SSS_ID_TYPE_UID);
-    PyModule_AddIntConstant(module, "ID_GROUP", SSS_ID_TYPE_GID);
-    PyModule_AddIntConstant(module, "ID_BOTH", SSS_ID_TYPE_BOTH);
-
-    PyModule_AddStringConstant(module, "SID_KEY", SSS_SID_KEY);
-    PyModule_AddStringConstant(module, "NAME_KEY", SSS_NAME_KEY);
-    PyModule_AddStringConstant(module, "ID_KEY", SSS_ID_KEY);
-    PyModule_AddStringConstant(module, "TYPE_KEY", SSS_TYPE_KEY);
+    if (ret == -1) MODINITERROR;
```

Comment:
Don't we need to cleanup here (dispose module)?
Otherwise looks like mem leak.

Reply:
I've had to check since I've never done any C code interact with Python. Apparently, it's not necessary since there are reference counts [1]. Once the local variable goes out of scope, the reference number is decremented and if this reference number is equal to zero then the address is freed.

Link:
[1] https://docs.python.org/3/c-api/intro.html#objects-types-and-reference-counts

Reply:
> Once the local variable goes out of scope

What variable do you mean?

Seems it's rather "not created" in this case...

Not sure I fully understand https://www.python.org/dev/peps/pep-3121/#id5 "Module Finalization". But since this module doesn't define `freefunc` this perhaps doesn't matter.

Reply:
> > Once the local variable goes out of scope
> 
> What variable do you mean?

The variable called ```module```, I thought you were speaking about it.

Now I see that you were speaking about ```pysss_nss_idmap_def```. Both ```m_free``` and ```m_clear``` aren't defined, contrary to what defines the web page that you have sent. So, I don't know what to say.

Reply:
> > > Once the local variable goes out of scope
> > 
> > 
> > What variable do you mean?
> 
> The variable called `module`

No, C will not do anything when variable goes out of scope (so no reference counting happens this moment).

But perhaps Python interpreter will handle this on its own (since init function didn't return an object). I don't know.

But again, since module doesn't have a "destructor", I guess it doesn't matter.

Reply:
Bump: Is there any resolution?

Reply:
Hi @ikerexxe ,

looking at the [example](https://github.com/python/cpython/blob/5c7ed7550ec2da16d7679e538fcd7c1a5631811f/Doc/includes/custom.c#L40), I still think `PyObject *module` needs explicit disposal in this case.

Perhaps, other option could be using [`Py_UNREACHABLE`](https://docs.python.org/3/c-api/intro.html#c.Py_UNREACHABLE) but this depends on module usage.

Reply:
Btw, I guess this means `PyInit_pyhbac()` and `PyInit_pysss()` handle those errors wrongly as well.

Reply:
After rereading [the documentation](https://docs.python.org/3/c-api/intro.html#objects-types-and-reference-counts) and your [example](https://github.com/python/cpython/blob/5c7ed7550ec2da16d7679e538fcd7c1a5631811f/Doc/includes/custom.c#L40), I agree with you that `PyObject *module` should be decremented if `PyModule_AddIntConstant` fails. Let me correct it and propose a new patch for this PR.

Regarding the other two modules that you mention, I could do it in this PR or create a new one.

Reply:
I would propose to cover this in a single PR because IMO this is really the same material.